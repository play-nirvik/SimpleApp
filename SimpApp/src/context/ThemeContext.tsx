import EncryptedStorage from 'react-native-encrypted-storage';
import React, {createContext, useContext, useEffect, useState} from 'react';

export const ThemeContext = createContext<ThemeProps>({});

export const useTheme = () => {
  return useContext(ThemeContext);
};

export const ThemeProvider = ({children}: any) => {
  const [theme, setTheme] = useState<string | null | undefined>('default');

  useEffect(() => {
    const loadTheme = async () => {
      const themeCountry = await EncryptedStorage.getItem('theme');

      if (themeCountry) {
        setTheme(themeCountry);
      }
    };

    loadTheme();
  }, []);

  const bgColorMap = () => {
    let bgColor = 'transparent';
    switch (theme) {
      case 'ae':
        bgColor = '#81c784';
        break;

      case 'in':
        bgColor = '#64b5f6';
        break;

      case 'fr':
        bgColor = '#7986cb';
        break;

      case 'es':
        bgColor = '#e57373';
        break;

      default:
        bgColor = 'transparent';
        break;
    }
    return bgColor;
  };

  const textColorMap = () => {
    let textColor = '#111';
    switch (theme) {
      case 'ae':
        textColor = '#f1f8e9';
        break;

      case 'in':
        textColor = '#e1f5fe';
        break;

      case 'fr':
        textColor = '#e8eaf6';
        break;

      case 'es':
        textColor = '#ffebee';
        break;

      default:
        textColor = '#111';
        break;
    }
    return textColor;
  };

  const updateHandler = async () => {
    const themeCountry = await EncryptedStorage.getItem('theme');

    setTheme(themeCountry);
  };

  const value = {
    onUpdate: updateHandler,
    bgColor: bgColorMap(),
    textColor: textColorMap(),
    theme,
    setTheme,
  };

  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
};
