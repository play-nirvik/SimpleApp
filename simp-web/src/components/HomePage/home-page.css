.body-content {
    display: flex;
    height: 85vh;
    padding: 1rem;
    flex-direction: column;
}

.langSelector {
    flex: 1;
}

.langSelector .formlabel {
    width: auto;
}