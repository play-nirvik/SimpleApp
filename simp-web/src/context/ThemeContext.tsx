/* eslint-disable react-refresh/only-export-components */
/* eslint-disable @typescript-eslint/no-explicit-any */
import secureLocalStorage from 'react-secure-storage';
import React, { createContext, useContext, useEffect, useState } from 'react';

export const ThemeContext = createContext<ThemeProps>({});

export const useTheme = () => {
  return useContext(ThemeContext);
};

export const ThemeProvider = ({ children }: any) => {
  const [theme, setTheme] = useState<string | null | undefined>('default');

  useEffect(() => {
    const themeCountry = secureLocalStorage.getItem('theme') as string;

    if (themeCountry) {
      setTheme(themeCountry);
    }
  }, []);

  const bgColorMap = () => {
    let bgColor = 'transparent';
    switch (theme) {
      case 'ae':
        bgColor = '#81c784';
        break;

      case 'in':
        bgColor = '#64b5f6';
        break;

      case 'fr':
        bgColor = '#7986cb';
        break;

      case 'es':
        bgColor = '#e57373';
        break;

      default:
        bgColor = 'transparent';
        break;
    }
    return bgColor;
  };

  const textColorMap = () => {
    let textColor = '#111';
    switch (theme) {
      case 'ae':
        textColor = '#f1f8e9';
        break;

      case 'in':
        textColor = '#e1f5fe';
        break;

      case 'fr':
        textColor = '#e8eaf6';
        break;

      case 'es':
        textColor = '#ffebee';
        break;

      default:
        textColor = '#111';
        break;
    }
    return textColor;
  };

  const updateHandler = async () => {
    const themeCountry = secureLocalStorage.getItem('theme') as string;

    setTheme(themeCountry);
  };

  const value = {
    onUpdate: updateHandler,
    bgColor: bgColorMap(),
    textColor: textColorMap(),
    theme,
    setTheme,
  };

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
};
