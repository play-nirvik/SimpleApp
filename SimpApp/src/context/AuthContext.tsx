import AsyncStorage from '@react-native-community/async-storage';
import axios from 'axios';
import React, {createContext, useContext, useEffect, useState} from 'react';
import {API_URL} from '../config/constants';
import {useTheme} from './ThemeContext';
import UserService from '../services/UserService';

const AuthContext = createContext<AuthProps>({});

export const useAuth = () => {
  return useContext(AuthContext);
};

export const AuthProvider = ({children}: any) => {
  const [isLoading, setIsLoading] = useState(true);
  const [authState, setAuthState] = useState<{
    token: string | null;
    authenticated: boolean | null;
    username: string | null;
  }>({token: null, authenticated: null, username: null});

  const {setTheme} = useTheme();

  useEffect(() => {
    const loadToken = async () => {
      const userInfo: UserInfo | undefined = await UserService.get();

      if (userInfo?.token) {
        setAuthState({
          token: userInfo?.token,
          authenticated: true,
          username: userInfo?.username,
        });
      }
      setIsLoading(false);
    };

    loadToken();
  }, []);

  const registerHandler = async (userData: UserProps) => {
    try {
      const result = await axios.post(`${API_URL}/register`, userData);

      return result;
    } catch (e) {
      return {error: true, message: (e as any).response.data.message};
    }
  };

  const loginHandler = async (username: string, password: string) => {
    try {
      const result = await axios.post(`${API_URL}/login`, {username, password});

      setAuthState({
        token: result.data.token,
        authenticated: true,
        username: result.data.username,
      });

      UserService.set(result.data);
      await AsyncStorage.setItem('theme', result.data.country);
      setTheme(result.data.country);

      return result;
    } catch (e) {
      console.log((e as any).response.data.message);
      return {error: true, msg: (e as any).response.data.message};
    }
  };

  const logoutHandler = async () => {
    UserService.delete();
    await AsyncStorage.removeItem('theme');

    setAuthState({
      token: null,
      authenticated: false,
      username: null,
    });
  };

  const value = {
    onRegister: registerHandler,
    onLogin: loginHandler,
    onLogout: logoutHandler,
    authState,
    isLoading,
  };
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};
