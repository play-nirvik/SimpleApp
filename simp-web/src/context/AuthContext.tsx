/* eslint-disable react-refresh/only-export-components */
/* eslint-disable @typescript-eslint/no-explicit-any */
import axios from 'axios';
import secureLocalStorage from 'react-secure-storage';
import React, { createContext, useContext, useEffect, useState } from 'react';
import { API_URL } from '../config/constants';
import { useTheme } from './ThemeContext';
import UserService from '../services/UserService';

const AuthContext = createContext<AuthProps>({});

export const useAuth = () => {
  return useContext(AuthContext);
};

export const AuthProvider = ({ children }: any) => {
  const [isLoading, setIsLoading] = useState(true);
  const [authState, setAuthState] = useState<{
    token: string | null;
    authenticated: boolean | null;
    username: string | null;
  }>({ token: null, authenticated: null, username: null });

  const { setTheme } = useTheme();

  useEffect(() => {
    const userInfo: UserInfo | undefined = UserService.get();

    if (userInfo?.token) {
      setAuthState({
        token: userInfo?.token,
        authenticated: true,
        username: userInfo?.username,
      });
    }
    setIsLoading(false);
  }, []);

  const registerHandler = async (userData: UserProps) => {
    try {
      const result = await axios.post(`${API_URL}/register`, userData);

      return result;
    } catch (e) {
      return { error: true, message: (e as any).response.data.message };
    }
  };

  const loginHandler = async (username: string, password: string) => {
    try {
      const result = await axios.post(`${API_URL}/login`, { username, password });

      console.log('login response>>', result.data);

      setAuthState({
        token: result?.data?.token,
        authenticated: true,
        username: result?.data?.username,
      });

      UserService.set(result?.data);
      secureLocalStorage.setItem('theme', result?.data?.country);
      setTheme(result?.data?.country);

      return result;
    } catch (e) {
      console.log((e as any).response.data.message);
      return { error: true, msg: (e as any).response.data.message };
    }
  };

  const logoutHandler = () => {
    UserService.delete();
    secureLocalStorage.removeItem('theme');

    setAuthState({
      token: null,
      authenticated: false,
      username: null,
    });
  };

  const value = {
    onRegister: registerHandler,
    onLogin: loginHandler,
    onLogout: logoutHandler,
    authState,
    isLoading,
  };
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};
