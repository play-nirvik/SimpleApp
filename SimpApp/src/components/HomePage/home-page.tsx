import React, {useCallback, useEffect, useState} from 'react';
import {StyleSheet, View, Text, Pressable} from 'react-native';
import {useTranslation} from 'react-i18next';
import {SUPPORTED_LANGUAGES} from '../../config/constants';
import DropDownPicker from 'react-native-dropdown-picker';
import {useTheme} from '../../context/ThemeContext';
import NotificationService from '../../services/NotificationService';
import {useAuth} from '../../context/AuthContext';

export default function HomePage(): React.JSX.Element {
  const {t, i18n} = useTranslation();
  const {bgColor, textColor} = useTheme();
  const {authState} = useAuth();

  const [open, setOpen] = useState(false);
  const [language, setLanguage] = useState<string>('en');
  const [items, setItems] = useState(SUPPORTED_LANGUAGES);

  const changeLanguage = useCallback(
    (value: string) => {
      i18n
        .changeLanguage(value)
        .then(() => setLanguage(value))
        .catch(err => console.log(err));
    },
    [i18n],
  );

  useEffect(() => {
    changeLanguage(language);
  }, [changeLanguage, language]);

  return (
    <View style={[styles.container, {backgroundColor: bgColor}]}>
      <Text style={[styles.styledText, {color: textColor}]}>
        {t('greeting')}, {authState?.username}
      </Text>

      <Pressable
        onPress={NotificationService.sendTestMessage}
        style={styles.pushButton}
        hitSlop={10}>
        <Text>{t('push.title')}</Text>
      </Pressable>

      <DropDownPicker
        open={open}
        value={language}
        items={items}
        setOpen={setOpen}
        setValue={setLanguage}
        setItems={setItems}
        style={styles.languageDropdown}
        listMode="MODAL"
        testID="languageDropdownButton"
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  styledText: {
    flex: 1,
    fontSize: 40,
    color: 'green',
    padding: 20,
  },
  pushButton: {
    position: 'absolute',
    top: '20%',
    left: 20,
  },
  languageDropdown: {
    width: '40%',
    alignSelf: 'center',
    justifyContent: 'flex-end',
    marginBottom: 20,
    flexDirection: 'row',
    verticalAlign: 'middle',
  },
});
