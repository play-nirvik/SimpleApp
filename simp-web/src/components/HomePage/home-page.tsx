import { SUPPORTED_LANGUAGES } from '../../config/constants';
import { useTheme } from '../../context/ThemeContext';
import UserService from '../../services/UserService';
import './home-page.css';
import React, { useCallback, useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import NotificationService from '../../services/NotificationService';
import toast, { Toaster } from 'react-hot-toast';

const HomePage: React.FC = () => {
  const navigate = useNavigate();
  const { bgColor, textColor } = useTheme();
  const { t, i18n } = useTranslation();

  const [displayName, setDisplayName] = useState('');
  const [language, setLanguage] = useState<string>('en');

  useEffect(() => {
    const userInfo = UserService.get();
    if (!userInfo || userInfo === undefined) {
      navigate('/');
    } else {
      setDisplayName(userInfo?.username);
    }
  }, [navigate]);

  NotificationService.onMessageListener()
    .then((payload) => {
      if (payload) {
        toast(t('push.content'));
      }
    })
    .catch((err) => console.log('failed: ', err));

  useEffect(() => {
    NotificationService.requestForToken();
  }, []);

  const changeLanguage = useCallback(
    (value: string) => {
      i18n
        .changeLanguage(value)
        .then(() => setLanguage(value))
        .catch((err) => console.log(err));
      console.log(value);
    },
    [i18n]
  );

  useEffect(() => {
    changeLanguage(language);
  }, [changeLanguage, language]);

  return (
    <>
      <Toaster />

      <div className='body-content' style={{ backgroundColor: bgColor }}>
        <p className='body-text' style={{ color: textColor }}>
          {t('greeting')}, {displayName}
        </p>

        <div>
          <button onClick={NotificationService.sendTestMessage}>{t('push.title')}</button>
        </div>

        <div className='langSelector'>
          <label>
            <span className='formlabel'>Select a language:</span>
            <select onChange={(e) => setLanguage(e.target.value)} className='fieldselect' defaultValue={language}>
              {SUPPORTED_LANGUAGES.map((option, index) => {
                return (
                  <option key={index} value={option.value}>
                    {option.label}
                  </option>
                );
              })}
            </select>
          </label>
        </div>
      </div>
    </>
  );
};

export default HomePage;
