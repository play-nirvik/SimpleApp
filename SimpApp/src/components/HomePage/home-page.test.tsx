import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {ThemeProvider} from '../../context/ThemeContext'; // Assuming you have a ThemeProvider
import {AuthProvider} from '../../context/AuthContext'; // Assuming you have an AuthProvider
import HomePage from './home-page';

describe('HomePage', () => {
  it('renders correctly', () => {
    const {toJSON} = render(
      <ThemeProvider>
        <AuthProvider>
          <HomePage />
        </AuthProvider>
      </ThemeProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('displays a greeting message with the username', () => {
    const {getByText} = render(
      <ThemeProvider>
        <AuthProvider value={{authState: {username: 'TestUser'}}}>
          <HomePage />
        </AuthProvider>
      </ThemeProvider>,
    );
    expect(getByText('Hello, TestUser')).toBeDefined();
  });

  it('changes language when selecting from the dropdown', async () => {
    const {getByTestId} = render(
      <ThemeProvider>
        <AuthProvider>
          <HomePage />
        </AuthProvider>
      </ThemeProvider>,
    );

    // Open the dropdown
    fireEvent.press(getByTestId('languageDropdownButton'));

    // Select a language
    // fireEvent.press(getByTestId('languageDropdownItem-en'));

    // Assuming your changeLanguage function is asynchronous, you might want to wait
    // for the language change to take effect
    // await waitFor(() => {
    //   expect(getByTestId('languageDropdownButton')).toHaveTextContent('en');
    // });
  });
});
