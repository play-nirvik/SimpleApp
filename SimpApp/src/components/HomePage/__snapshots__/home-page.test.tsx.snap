// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders correctly 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "transparent",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "green",
          "flex": 1,
          "fontSize": 40,
          "padding": 20,
        },
        {
          "color": "#111",
        },
      ]
    }
  >
    greeting
    , 
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={10}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "left": 20,
        "position": "absolute",
        "top": "20%",
      }
    }
  >
    <Text>
      push.title
    </Text>
  </View>
  <View
    style={
      [
        {
          "width": "100%",
        },
        {
          "zIndex": 5000,
        },
        {},
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#fff",
          "borderColor": "#000",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "marginBottom": 20,
          "minHeight": 50,
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 3,
          "verticalAlign": "middle",
          "width": "40%",
          "zIndex": 5000,
        }
      }
      testID="languageDropdownButton"
    >
      <Text
        style={
          [
            {
              "color": "#000",
              "flex": 1,
            },
            {},
            {},
            false,
          ]
        }
      >
        English
      </Text>
      <View
        style={
          [
            {
              "marginLeft": 10,
            },
            {},
          ]
        }
      >
        <Image
          source={
            {
              "testUri": "../../react-native-dropdown-picker/src/themes/light/icons/arrow-down.png",
            }
          }
          style={
            [
              {
                "height": 20,
                "width": 20,
              },
              {},
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
